// @pkmn/sim/build/cjs/sim/dex-moves.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DexMoves = exports.DataMove = void 0;
const utils_1 = require("../lib/utils");
const dex_data_1 = require("./dex-data");
class DataMove extends dex_data_1.BasicEffect {
    constructor(data) {
        super(data);
        this.fullname = `move: ${this.name}`;
        this.effectType = 'Move';
        this.type = utils_1.Utils.getString(data.type);
        this.target = data.target;
        this.basePower = Number(data.basePower);
        this.accuracy = data.accuracy;
        this.critRatio = Number(data.critRatio) || 1;
        this.baseMoveType = utils_1.Utils.getString(data.baseMoveType) || this.type;
        this.secondary = data.secondary || null;
        this.secondaries = data.secondaries || (this.secondary && [this.secondary]) || null;
        this.hasSheerForce = !!(data.hasSheerForce && !this.secondaries);
        this.priority = Number(data.priority) || 0;
        this.category = data.category;
        this.overrideOffensiveStat = data.overrideOffensiveStat || undefined;
        this.overrideOffensivePokemon = data.overrideOffensivePokemon || undefined;
        this.overrideDefensiveStat = data.overrideDefensiveStat || undefined;
        this.overrideDefensivePokemon = data.overrideDefensivePokemon || undefined;
        this.ignoreNegativeOffensive = !!data.ignoreNegativeOffensive;
        this.ignorePositiveDefensive = !!data.ignorePositiveDefensive;
        this.ignoreOffensive = !!data.ignoreOffensive;
        this.ignoreDefensive = !!data.ignoreDefensive;
        this.ignoreImmunity = (data.ignoreImmunity !== undefined ? data.ignoreImmunity : this.category === 'Status');
        this.pp = Number(data.pp);
        this.noPPBoosts = !!(data.noPPBoosts ?? data.isZ);
        this.isZ = data.isZ || false;
        this.isMax = data.isMax || false;
        this.flags = data.flags || {};
        this.selfSwitch = (typeof data.selfSwitch === 'string' ? data.selfSwitch : data.selfSwitch) || undefined;
        this.nonGhostTarget = data.nonGhostTarget || '';
        this.ignoreAbility = data.ignoreAbility || false;
        this.damage = data.damage;
        this.spreadHit = data.spreadHit || false;
        this.forceSTAB = !!data.forceSTAB;
        this.volatileStatus = typeof data.volatileStatus === 'string' ? data.volatileStatus : undefined;
        if (this.category !== 'Status' && !data.maxMove && this.id !== 'struggle') {
            this.maxMove = { basePower: 1 };
            if (this.isMax || this.isZ) {
                // already initialized to 1
            }
            else if (!this.basePower) {
                this.maxMove.basePower = 100;
            }
            else if (['Fighting', 'Poison'].includes(this.type)) {
                if (this.basePower >= 150) {
                    this.maxMove.basePower = 100;
                }
                else if (this.basePower >= 110) {
                    this.maxMove.basePower = 95;
                }
                else if (this.basePower >= 75) {
                    this.maxMove.basePower = 90;
                }
                else if (this.basePower >= 65) {
                    this.maxMove.basePower = 85;
                }
                else if (this.basePower >= 55) {
                    this.maxMove.basePower = 80;
                }
                else if (this.basePower >= 45) {
                    this.maxMove.basePower = 75;
                }
                else {
                    this.maxMove.basePower = 70;
                }
            }
            else {
                if (this.basePower >= 150) {
                    this.maxMove.basePower = 150;
                }
                else if (this.basePower >= 110) {
                    this.maxMove.basePower = 140;
                }
                else if (this.basePower >= 75) {
                    this.maxMove.basePower = 130;
                }
                else if (this.basePower >= 65) {
                    this.maxMove.basePower = 120;
                }
                else if (this.basePower >= 55) {
                    this.maxMove.basePower = 110;
                }
                else if (this.basePower >= 45) {
                    this.maxMove.basePower = 100;
                }
                else {
                    this.maxMove.basePower = 90;
                }
            }
        }
        if (this.category !== 'Status' && !data.zMove && !this.isZ && !this.isMax && this.id !== 'struggle') {
            let basePower = this.basePower;
            this.zMove = {};
            if (Array.isArray(data.multihit))
                basePower *= 3;
            if (!basePower) {
                this.zMove.basePower = 100;
            }
            else if (basePower >= 140) {
                this.zMove.basePower = 200;
            }
            else if (basePower >= 130) {
                this.zMove.basePower = 195;
            }
            else if (basePower >= 120) {
                this.zMove.basePower = 190;
            }
            else if (basePower >= 110) {
                this.zMove.basePower = 185;
            }
            else if (basePower >= 100) {
                this.zMove.basePower = 180;
            }
            else if (basePower >= 90) {
                this.zMove.basePower = 175;
            }
            else if (basePower >= 80) {
                this.zMove.basePower = 160;
            }
            else if (basePower >= 70) {
                this.zMove.basePower = 140;
            }
            else if (basePower >= 60) {
                this.zMove.basePower = 120;
            }
            else {
                this.zMove.basePower = 100;
            }
        }
        if (!this.gen) {
            // special handling for gen8 gmax moves (all of them have num 1000 but they are part of gen8)
            if (this.num >= 827 && !this.isMax) {
                this.gen = 9;
            }
            else if (this.num >= 743) {
                this.gen = 8;
            }
            else if (this.num >= 622) {
                this.gen = 7;
            }
            else if (this.num >= 560) {
                this.gen = 6;
            }
            else if (this.num >= 468) {
                this.gen = 5;
            }
            else if (this.num >= 355) {
                this.gen = 4;
            }
            else if (this.num >= 252) {
                this.gen = 3;
            }
            else if (this.num >= 166) {
                this.gen = 2;
            }
            else if (this.num >= 1) {
                this.gen = 1;
            }
        }
        (0, dex_data_1.assignMissingFields)(this, data);
    }
}
exports.DataMove = DataMove;
const EMPTY_MOVE = utils_1.Utils.deepFreeze(new DataMove({ name: '', exists: false }));
class DexMoves {
    constructor(dex) {
        this.moveCache = new Map();
        this.allCache = null;
        this.dex = dex;
    }
    get(name) {
        if (name && typeof name !== 'string')
            return name;
        const id = name ? (0, dex_data_1.toID)(name.trim()) : '';
        return this.getByID(id);
    }
    getByID(id) {
        if (id === '' || id === 'constructor')
            return EMPTY_MOVE;
        let move = this.moveCache.get(id);
        if (move)
            return move;
        if (this.dex.getAlias(id)) {
            move = this.get(this.dex.getAlias(id));
            if (move.exists) {
                this.moveCache.set(id, move);
            }
            return move;
        }
        if (id.startsWith('hiddenpower')) {
            id = /([a-z]*)([0-9]*)/.exec(id)[1];
        }
        if (id && this.dex.data.Moves.hasOwnProperty(id)) {
            const moveData = this.dex.data.Moves[id];
            const moveTextData = this.dex.getDescs('Moves', id, moveData);
            move = new DataMove({
                name: id,
                ...moveData,
                ...moveTextData,
            });
            if (move.gen > this.dex.gen) {
                move.isNonstandard = 'Future';
            }
            if (this.dex.parentMod) {
                // If move is exactly identical to parentMod's move, reuse parentMod's copy
                const parentMod = this.dex.mod(this.dex.parentMod);
                if (moveData === parentMod.data.Moves[id]) {
                    const parentMove = parentMod.moves.getByID(id);
                    if (move.isNonstandard === parentMove.isNonstandard &&
                        move.desc === parentMove.desc && move.shortDesc === parentMove.shortDesc) {
                        move = parentMove;
                    }
                }
            }
        }
        else {
            move = new DataMove({
                name: id, exists: false,
            });
        }
        move.kind = 'Move';
        if (move.exists)
            this.moveCache.set(id, move);
        return move;
    }
    all() {
        if (this.allCache)
            return this.allCache;
        const moves = [];
        for (const id in this.dex.data.Moves) {
            moves.push(this.getByID(id));
        }
        this.allCache = moves;
        return this.allCache;
    }
}
exports.DexMoves = DexMoves;
