// @pkmn/sim/build/cjs/sim/dex.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleTable = exports.Dex = exports.ModdedDex = exports.toID = void 0;
const Utils = __importStar(require("../lib/utils"));
const dex_conditions_1 = require("./dex-conditions");
const dex_moves_1 = require("./dex-moves");
const dex_items_1 = require("./dex-items");
const dex_abilities_1 = require("./dex-abilities");
const dex_species_1 = require("./dex-species");
const dex_formats_1 = require("./dex-formats");
Object.defineProperty(exports, "RuleTable", { enumerable: true, get: function () { return dex_formats_1.RuleTable; } });
const aliases_1 = require("../data/aliases");
const gen1 = __importStar(require("../data/mods/gen1"));
const gen2 = __importStar(require("../data/mods/gen2"));
const gen3 = __importStar(require("../data/mods/gen3"));
const gen4 = __importStar(require("../data/mods/gen4"));
const gen5 = __importStar(require("../data/mods/gen5"));
const gen6 = __importStar(require("../data/mods/gen6"));
const gen7 = __importStar(require("../data/mods/gen7"));
const gen8 = __importStar(require("../data/mods/gen8"));
const gen8bdsp = __importStar(require("../data/mods/gen8bdsp"));
const gen8legends = __importStar(require("../data/mods/gen8legends"));
const gen9 = __importStar(require("../data"));
const formats9 = __importStar(require("../data/formats-data"));
const learn2 = __importStar(require("../data/mods/gen2/learnsets"));
const learn6 = __importStar(require("../data/mods/gen6/learnsets"));
const learn8 = __importStar(require("../data/mods/gen8/learnsets"));
const learn8bdsp = __importStar(require("../data/mods/gen8bdsp/learnsets"));
const learn8legends = __importStar(require("../data/mods/gen8legends/learnsets"));
const learn9 = __importStar(require("../data/learnsets"));
const pokemongo = __importStar(require("../data/pokemongo"));
const legality2 = __importStar(require("../data/mods/gen2/legality"));
const legality6 = __importStar(require("../data/mods/gen6/legality"));
const legality8 = __importStar(require("../data/mods/gen8/legality"));
const legality8bdsp = __importStar(require("../data/mods/gen8bdsp/legality"));
const legality8legends = __importStar(require("../data/mods/gen8legends/legality"));
const legality9 = __importStar(require("../data/legality"));
const abilities_1 = require("../data/text/abilities");
const items_1 = require("../data/text/items");
const moves_1 = require("../data/text/moves");
const default_1 = require("../data/text/default");
const Data = __importStar(require("./dex-data"));
const BASE_MOD = 'gen9';
const LEARN_ORDER = 'MTLREVDSC';
function merge(learnsets, legality) {
    const merged = { ...learnsets };
    for (const id in legality.Legality) {
        for (const key in legality.Legality[id]) {
            if (key === 'learnset') {
                for (const moveid in legality.Legality[id][key]) {
                    const special = legality.Legality[id][key][moveid];
                    // import guarantees {id: {learnsets: {}}} exists in learnsets iff it exists in legality
                    const existing = merged.Learnsets[id][key];
                    if (existing[moveid]) {
                        existing[moveid].push(...special);
                        existing[moveid].sort((a, b) => +b.charAt(0) - +a.charAt(0) ||
                            LEARN_ORDER.indexOf(a.charAt(1)) - LEARN_ORDER.indexOf(b.charAt(1)));
                    }
                    else {
                        existing[moveid] = special;
                    }
                }
            }
            else {
                // @ts-ignore
                merged.Learnsets[id][key] = legality.Legality[id][key];
            }
        }
    }
    return merged;
}
const dexData = {
    gen1, gen2: { ...gen2, ...merge(learn2, legality2) }, gen3,
    gen4, gen5, gen6: { ...gen6, ...merge(learn6, legality6) },
    gen7, gen8: { ...gen8, ...merge(learn8, legality8) },
    gen8bdsp: { ...gen8bdsp, ...merge(learn8bdsp, legality8bdsp) },
    gen8legends: { ...gen8legends, ...merge(learn8legends, legality8legends) },
    gen9: { ...gen9, ...formats9, ...pokemongo, ...merge(learn9, legality9) },
};
const dexes = Object.create(null);
const DATA_TYPES = [
    'Abilities', 'Rulesets', 'FormatsData', 'Items', 'Learnsets', 'Moves',
    'Natures', 'Pokedex', 'Scripts', 'Conditions', 'TypeChart', 'PokemonGoData',
];
const TEXT = {
    Abilities: abilities_1.AbilitiesText,
    Items: items_1.ItemsText,
    Moves: moves_1.MovesText,
    // Pokedex: PokedexText as DexTable<PokedexText>,
    Default: default_1.DefaultText,
};
exports.toID = Data.toID;
class ModdedDex {
    constructor(mod = 'base') {
        this.Data = Data;
        this.Condition = dex_conditions_1.Condition;
        this.Ability = dex_abilities_1.Ability;
        this.Item = dex_items_1.Item;
        this.Move = dex_moves_1.DataMove;
        this.Species = dex_species_1.Species;
        this.Format = dex_formats_1.Format;
        this.ModdedDex = ModdedDex;
        this.name = "[ModdedDex]";
        this.toID = Data.toID;
        this.aliases = null;
        this.gen = 0;
        this.parentMod = '';
        this.modsLoaded = false;
        this.deepClone = Utils.deepClone;
        this.deepFreeze = Utils.deepFreeze;
        this.Multiset = Utils.Multiset;
        this.isBase = (mod === 'base');
        this.currentMod = mod;
        this.dataCache = null;
        this.formats = new dex_formats_1.DexFormats(this);
        this.abilities = new dex_abilities_1.DexAbilities(this);
        this.items = new dex_items_1.DexItems(this);
        this.moves = new dex_moves_1.DexMoves(this);
        this.species = new dex_species_1.DexSpecies(this);
        this.learnsets = new dex_species_1.DexLearnsets(this);
        this.conditions = new dex_conditions_1.DexConditions(this);
        this.natures = new Data.DexNatures(this);
        this.types = new Data.DexTypes(this);
        this.stats = new Data.DexStats(this);
    }
    get modid() {
        return this.currentMod;
    }
    get data() {
        return this.loadData();
    }
    get dexes() {
        return dexes;
    }
    mod(mod, modData) {
        if (!mod)
            return dexes['base'];
        const modid = (0, exports.toID)(mod);
        if (modData?.Types && !modData.TypeChart)
            modData.TypeChart = modData.Types;
        if (modData?.Species && !modData.Pokedex)
            modData.Pokedex = modData.Species;
        const dex = (modid in dexes) && !modData ? dexes[modid] : new ModdedDex(modid);
        dex.loadData(modData);
        if (!(modid in dexes))
            dexes[modid] = dex;
        if (modData?.Formats)
            dex.formats.extend(modData?.Formats);
        return dex;
    }
    forGen(gen) {
        if (gen < 1 || gen > 9)
            throw new Error(`Unsupported gen ${gen}`);
        return this.mod(`gen${gen}`);
    }
    forFormat(format) {
        const mod = this.formats.get(format).mod;
        return dexes[mod || BASE_MOD].includeData();
    }
    modData(dataType, id) {
        if (dataType === 'Scripts')
            throw new Error(`'${dataType}' cannot be indexed by '${id}'`);
        if (this.isBase)
            return this.data[dataType][id];
        if (this.data[dataType][id] !== dexes[this.parentMod].data[dataType][id])
            return this.data[dataType][id];
        return (this.data[dataType][id] = Utils.deepClone(this.data[dataType][id]));
    }
    effectToString() {
        return this.name;
    }
    /**
     * Sanitizes a username or Pokemon nickname
     *
     * Returns the passed name, sanitized for safe use as a name in the PS
     * protocol.
     *
     * Such a string must uphold these guarantees:
     * - must not contain any ASCII whitespace character other than a space
     * - must not start or end with a space character
     * - must not contain any of: | , [ ]
     * - must not be the empty string
     * - must not contain Unicode RTL control characters
     *
     * If no such string can be found, returns the empty string. Calling
     * functions are expected to check for that condition and deal with it
     * accordingly.
     *
     * getName also enforces that there are not multiple consecutive space
     * characters in the name, although this is not strictly necessary for
     * safety.
     */
    getName(name) {
        if (typeof name !== 'string' && typeof name !== 'number')
            return '';
        name = ('' + name).replace(/[|\s[\],\u202e]+/g, ' ').trim();
        if (name.length > 18)
            name = name.substr(0, 18).trim();
        // remove zalgo
        name = name.replace(/[\u0300-\u036f\u0483-\u0489\u0610-\u0615\u064B-\u065F\u0670\u06D6-\u06DC\u06DF-\u06ED\u0E31\u0E34-\u0E3A\u0E47-\u0E4E]{3,}/g, '');
        name = name.replace(/[\u239b-\u23b9]/g, '');
        return name;
    }
    /**
     * Returns false if the target is immune; true otherwise.
     * Also checks immunity to some statuses.
     */
    getImmunity(source, target) {
        const sourceType = typeof source !== 'string' ? source.type : source;
        // @ts-ignore
        const targetTyping = target.getTypes?.() || target.types || target;
        if (Array.isArray(targetTyping)) {
            for (const type of targetTyping) {
                if (!this.getImmunity(sourceType, type))
                    return false;
            }
            return true;
        }
        const typeData = this.types.get(targetTyping);
        if (typeData && typeData.damageTaken[sourceType] === 3)
            return false;
        return true;
    }
    getEffectiveness(source, target) {
        const sourceType = typeof source !== 'string' ? source.type : source;
        // @ts-ignore
        const targetTyping = target.getTypes?.() || target.types || target;
        let totalTypeMod = 0;
        if (Array.isArray(targetTyping)) {
            for (const type of targetTyping) {
                totalTypeMod += this.getEffectiveness(sourceType, type);
            }
            return totalTypeMod;
        }
        const typeData = this.types.get(targetTyping);
        if (!typeData)
            return 0;
        switch (typeData.damageTaken[sourceType]) {
            case 1: return 1; // super-effective
            case 2: return -1; // resist
            // in case of weird situations like Gravity, immunity is handled elsewhere
            default: return 0;
        }
    }
    getDescs(table, id, dataEntry) {
        if (dataEntry.shortDesc) {
            return {
                desc: dataEntry.desc,
                shortDesc: dataEntry.shortDesc,
            };
        }
        const entry = TEXT[table][id];
        if (!entry)
            return null;
        const descs = {
            desc: '',
            shortDesc: '',
        };
        for (let i = this.gen; i < dexes['base'].gen; i++) {
            const curDesc = entry[`gen${i}`]?.desc;
            const curShortDesc = entry[`gen${i}`]?.shortDesc;
            if (!descs.desc && curDesc) {
                descs.desc = curDesc;
            }
            if (!descs.shortDesc && curShortDesc) {
                descs.shortDesc = curShortDesc;
            }
            if (descs.desc && descs.shortDesc)
                break;
        }
        if (!descs.shortDesc)
            descs.shortDesc = entry.shortDesc || '';
        if (!descs.desc)
            descs.desc = entry.desc || descs.shortDesc;
        return descs;
    }
    /**
     * Ensure we're working on a copy of a move (and make a copy if we aren't)
     *
     * Remember: "ensure" - by default, it won't make a copy of a copy:
     *     moveCopy === Dex.getActiveMove(moveCopy)
     *
     * If you really want to, use:
     *     moveCopyCopy = Dex.getActiveMove(moveCopy.id)
     */
    getActiveMove(move) {
        if (move && typeof move.hit === 'number')
            return move;
        move = this.moves.get(move);
        const moveCopy = Utils.deepClone(move);
        moveCopy.hit = 0;
        return moveCopy;
    }
    getHiddenPower(ivs) {
        const hpTypes = [
            'Fighting', 'Flying', 'Poison', 'Ground', 'Rock', 'Bug', 'Ghost', 'Steel',
            'Fire', 'Water', 'Grass', 'Electric', 'Psychic', 'Ice', 'Dragon', 'Dark',
        ];
        const tr = this.trunc;
        const stats = { hp: 31, atk: 31, def: 31, spe: 31, spa: 31, spd: 31 };
        if (this.gen <= 2) {
            // Gen 2 specific Hidden Power check. IVs are still treated 0-31 so we get them 0-15
            const atkDV = tr(ivs.atk / 2);
            const defDV = tr(ivs.def / 2);
            const speDV = tr(ivs.spe / 2);
            const spcDV = tr(ivs.spa / 2);
            return {
                type: hpTypes[4 * (atkDV % 4) + (defDV % 4)],
                power: tr((5 * ((spcDV >> 3) + (2 * (speDV >> 3)) + (4 * (defDV >> 3)) + (8 * (atkDV >> 3))) + (spcDV % 4)) / 2 + 31),
            };
        }
        else {
            // Hidden Power check for Gen 3 onwards
            let hpTypeX = 0;
            let hpPowerX = 0;
            let i = 1;
            for (const s in stats) {
                hpTypeX += i * (ivs[s] % 2);
                hpPowerX += i * (tr(ivs[s] / 2) % 2);
                i *= 2;
            }
            return {
                type: hpTypes[tr(hpTypeX * 15 / 63)],
                // After Gen 6, Hidden Power is always 60 base power
                power: (this.gen && this.gen < 6) ? tr(hpPowerX * 40 / 63) + 30 : 60,
            };
        }
    }
    /**
     * Truncate a number into an unsigned 32-bit integer, for
     * compatibility with the cartridge games' math systems.
     */
    trunc(num, bits = 0) {
        if (bits)
            return (num >>> 0) % (2 ** bits);
        return num >>> 0;
    }
    loadDataFile(mod, dataType, modData) {
        if (modData)
            return modData[dataType];
        return dexData[mod === 'base' ? BASE_MOD : mod][dataType];
    }
    getAlias(id) {
        return this.loadAliases().get(id);
    }
    loadAliases() {
        if (!this.isBase)
            return exports.Dex.loadAliases();
        if (this.aliases)
            return this.aliases;
        const aliases = new Map();
        for (const [alias, target] of Object.entries(aliases_1.Aliases)) {
            aliases.set(alias, (0, exports.toID)(target));
        }
        this.aliases = aliases;
        return this.aliases;
    }
    includeMods() {
        return this;
    }
    includeModData() {
        for (const mod in this.dexes) {
            dexes[mod].includeData();
        }
        return this;
    }
    includeData() {
        this.loadData();
        return this;
    }
    loadData(modData) {
        if (this.dataCache)
            return this.dataCache;
        const dataCache = {};
        const Scripts = this.loadDataFile(this.currentMod, 'Scripts', modData) || {};
        // We want to inherit most of Scripts but not this.
        const init = Scripts.init;
        this.parentMod = this.isBase ? '' : (Scripts.inherit || 'base');
        let parentDex;
        if (this.parentMod) {
            parentDex = dexes[this.parentMod];
            if (!parentDex || parentDex === this) {
                throw new Error(`Unable to load ${this.currentMod}. 'inherit' should specify a parent mod from which to inherit data, or must be not specified.`);
            }
        }
        if (!parentDex) {
            // Formats are inherited by mods and used by Rulesets
            this.includeFormats();
        }
        for (const dataType of DATA_TYPES) {
            dataCache[dataType] = this.loadDataFile(this.currentMod, dataType, modData);
            if (dataType === 'Rulesets' && !parentDex) {
                for (const format of this.formats.all()) {
                    dataCache.Rulesets[format.id] = { ...format, ruleTable: null };
                }
            }
        }
        if (parentDex) {
            for (const dataType of DATA_TYPES) {
                const parentTypedData = parentDex.data[dataType];
                if (!dataCache[dataType] && !init) {
                    dataCache[dataType] = parentTypedData;
                    continue;
                }
                const childTypedData = dataCache[dataType] || (dataCache[dataType] = {});
                for (const entryId in parentTypedData) {
                    if (childTypedData[entryId] === null) {
                        // null means don't inherit
                        delete childTypedData[entryId];
                    }
                    else if (!(entryId in childTypedData)) {
                        // If it doesn't exist it's inherited from the parent data
                        childTypedData[entryId] = parentTypedData[entryId];
                    }
                    else if (childTypedData[entryId]?.inherit) {
                        // {inherit: true} can be used to modify only parts of the parent data,
                        // instead of overwriting entirely
                        delete childTypedData[entryId].inherit;
                        // Merge parent into children entry, preserving existing childs' properties.
                        childTypedData[entryId] = { ...parentTypedData[entryId], ...childTypedData[entryId] };
                    }
                }
            }
        }
        // Flag the generation. Required for team validator.
        this.gen = dataCache.Scripts.gen;
        if (!this.gen)
            throw new Error(`Mod ${this.currentMod} needs a generation number in scripts.js`);
        dataCache.Types = dataCache.TypeChart;
        dataCache.Species = dataCache.Pokedex;
        this.dataCache = dataCache;
        // Execute initialization script.
        if (init)
            init.call(this);
        return this.dataCache;
    }
    includeFormats() {
        this.formats.load();
        return this;
    }
}
exports.ModdedDex = ModdedDex;
dexes['base'] = new ModdedDex();
dexes['gen1'] = new ModdedDex('gen1');
dexes['gen2'] = new ModdedDex('gen2');
dexes['gen3'] = new ModdedDex('gen3');
dexes['gen4'] = new ModdedDex('gen4');
dexes['gen5'] = new ModdedDex('gen5');
dexes['gen6'] = new ModdedDex('gen6');
dexes['gen7'] = new ModdedDex('gen7');
dexes['gen8'] = new ModdedDex('gen8');
// required for gen 9 moveset validation
dexes['gen8bdsp'] = new ModdedDex('gen8bdsp');
dexes['gen8legends'] = new ModdedDex('gen8legends');
// "gen9" is an alias for the current base data
dexes[BASE_MOD] = dexes['base'];
dexes['base'].includeModData();
exports.Dex = dexes['base'];
