// @pkmn/streams/index.js
class AsyncQueue {
  constructor() {
    this.items = [];
    this.waiters = [];
    this.ended = false;
    this.error = null;
  }
  push(value) {
    if (this.ended) return;
    const waiter = this.waiters.shift();
    if (waiter) waiter.resolve({value, done: false});
    else this.items.push(value);
  }
  end() {
    if (this.ended) return;
    this.ended = true;
    while (this.waiters.length) this.waiters.shift().resolve({value: undefined, done: true});
  }
  fail(error) {
    this.error = error || new Error('Stream error');
    this.ended = true;
    while (this.waiters.length) this.waiters.shift().reject(this.error);
  }
  next() {
    if (this.items.length) return Promise.resolve({value: this.items.shift(), done: false});
    if (this.error) return Promise.reject(this.error);
    if (this.ended) return Promise.resolve({value: undefined, done: true});
    return new Promise((resolve, reject) => this.waiters.push({resolve, reject}));
  }
}
class ObjectReadStream {
  constructor(options = {}) {
    this._queue = new AsyncQueue();
    this._options = options;
  }
  push(value) { this._queue.push(value); }
  pushEnd() { this._queue.end(); }
  pushError(error) { this._queue.fail(error); }
  async read() { const out = await this._queue.next(); return out.done ? null : out.value; }
  [Symbol.asyncIterator]() { return this; }
  next() { return this._queue.next(); }
}
class ObjectWriteStream extends ObjectReadStream {
  constructor(options = {}) { super(options); }
  write(value) {
    if (typeof this._options.write === 'function') return Promise.resolve(this._options.write(value));
    if (typeof this._write === 'function') return Promise.resolve(this._write(value));
    return Promise.resolve();
  }
  destroy() {
    if (typeof this._options.destroy === 'function') this._options.destroy();
    this.pushEnd();
  }
}
class ObjectReadWriteStream extends ObjectWriteStream {}
class ReadStream extends ObjectReadStream {}
class WriteStream extends ObjectWriteStream {}
class ReadWriteStream extends ObjectReadWriteStream {}
module.exports = {AsyncQueue, ObjectReadStream, ObjectWriteStream, ObjectReadWriteStream, ReadStream, WriteStream, ReadWriteStream};
